src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/rewriting/rules/GraphPatternConstructor.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    public static SPARQLGraphPattern createNewGraphPatternWithAND(LogicalOpBGPAdd lopBGPAdd, LogicalOpRequest<?,?> lopReq) {
+        final BGP bgpOfBGPAdd = lopBGPAdd.getBGP();
+        final BasicPattern bgp = ( (OpBGP)bgpOfBGPAdd.asJenaOp() ).getPattern();
+        final ElementTriplesBlock elementBGP = new ElementTriplesBlock( bgp );
+
+        final SPARQLQuery queryOfReq = ((SPARQLRequest) lopReq.getRequest()).getQuery();
+        final Element elementPattern = queryOfReq.asJenaQuery().getQueryPattern();
+
+        ((ElementGroup) elementPattern).addElement(elementBGP);
+
+        return new SPARQLGraphPatternImpl(elementPattern);
+    }
+
+    public static SPARQLGraphPattern createNewGraphPatternWithAND( final LogicalOpRequest<?, ?> lopReq1, final LogicalOpRequest<?, ?> lopReq2 ) {
+        final SPARQLQuery graphPattern1 = ((SPARQLRequest) lopReq1.getRequest()).getQuery();
+        final Element element1 = graphPattern1.asJenaQuery().getQueryPattern();
+
+        final SPARQLQuery graphPattern2 = ((SPARQLRequest) lopReq2.getRequest()).getQuery();
+        final Element element2 = graphPattern2.asJenaQuery().getQueryPattern();
+
+        ((ElementGroup) element1).addElement(element2);
+
+        return new SPARQLGraphPatternImpl(element1);
+    }
+
+    public static SPARQLGraphPattern createNewGraphPatternWithUnion( final LogicalOpRequest<?, ?> lopReq1, final LogicalOpRequest<?, ?> lopReq2 ) {
+        final SPARQLQuery graphPattern1 = ((SPARQLRequest) lopReq1.getRequest()).getQuery();
+        final Element element1 = graphPattern1.asJenaQuery().getQueryPattern();
+
+        final SPARQLQuery graphPattern2 = ((SPARQLRequest) lopReq2.getRequest()).getQuery();
+        final Element element2 = graphPattern2.asJenaQuery().getQueryPattern();
+
+        final ElementUnion elementUnion = new ElementUnion(element1);
+        elementUnion.addElement(element2);
+
+        return new SPARQLGraphPatternImpl(elementUnion);
+    }
+
+    protected static Set<TriplePattern> getTriplePatternsOfReq( final LogicalOpRequest<?, ?> lop ) {
+        final DataRetrievalRequest req = lop.getRequest();
+
+        if ( req instanceof TriplePatternRequest) {
+            final TriplePatternRequest tpReq = (TriplePatternRequest) lop.getRequest();
+            final Set<TriplePattern> tps = new HashSet<>();
+            tps.add( tpReq.getQueryPattern() );
+
+            return tps;
+        }
+        else if ( req instanceof BGPRequest) {
+            final BGPRequest bgpReq = (BGPRequest) lop.getRequest();
+            final BGP bgp = bgpReq.getQueryPattern();
+
+            if ( bgp.getTriplePatterns().size() == 0 ) {
+                throw new IllegalArgumentException( "the BGP is empty" );
+            }
+            else {
+                return new HashSet<>( bgp.getTriplePatterns() );
+            }
+        }
+        else  {
+            throw new IllegalArgumentException( "Cannot get triple patterns of the req" );
```

Comment:
```suggestion
            throw new IllegalArgumentException( "Cannot get triple patterns of the given request operator (type: " + req.getClass().getName() + ")." );
```
It may be helpful to have some more information if this exception will occur in some unexpected circumstances later ;-)